### PR TITLE
Multiextruder Machine Settings / Custom Multiextruder printer

### PR DESCRIPTION
**Pull request by fieldOfView**
Machine Settings currently does not support multi-extruder printers. It is also not possible to set the number of extruders.

The main difficulty of setting the number of extruders is that each extruder has its own definition, and these definitions need to have a back-reference to the main definition.

~~This PR is a proof of concept where it is possible to change the number of extruders limited to the number of extruders that have their own definition. So basically it lets the user turn off extruders (eg make a single extruder machine out of a UM3). While this may seem very limiting, the PR also defines 8 extruders for the Custom FDM Printer definition, so it is possible to create a custom printer with up to 8 extruders (though it still defaults to a single extruder machine).~~

~~Adding settings for the extruders (most notably extruder offset and nozzle switch gcode) is next up for this PR.~~

**Update:**
What this PR adds:
* A dropdown field for the number of extruders for the printer, limited to the number of extruders defined in the machine definition
* Up to 8 extruders for the Custom FDM Printer
* Tabs per extruder with settings for nozzle size, nozzle offsets and extruder start & end gcode for any printer with > 1 extruder

What this PR does not add:
* The ability to add extruders to definitions that do not add extruders (eg it is not possible with this PR to change the Ultimaker 2 definition into a dual extrusion machine)
* Nozzle size fields if the definition has variants (because the variant would override the definition_changes anyway)

### CODE REVIEW

**Comment by fieldOfView** [comment]
Any hints on why #042f45f cannot be built?

**Comment by awhiemstra** [comment]
Seems it cannot be built because it does not exist. O_o

I do not really know what's going on here. I guess Jenkins initial tried building it but now thinks it no longer exists?

**Comment by maukcc** [comment]
Will this only affect the custom printer, or can we implement this in other "definition".def.json files?
I would love it, if extruders in the side bar are greyed out to show that they are not being used.
And if they are not being used, they are not being heated up, like we are discussing in #1296 (regardless of what material was selected for it)

**Comment by Ghostkeeper** [comment]
The menu to adjust machine settings can be added to any printer definition file. It'll show the menu after the user chooses to add the printer and it'll show a button in the machine manager to restore the window later.

Just add this code to the definition metadata: https://github.com/Ultimaker/Cura/blob/master/resources/definitions/custom.def.json#L13

This pull request would make it possible to adjust extruder settings in multiple extruders with that window as well.

**Comment by fieldOfView** [comment]
> Just add this code to the definition metadata: https://github.com/Ultimaker/Cura/blob/master/resources/definitions/custom.def.json#L13

Only do that if you want the machine settings to pop up directly when adding the printer. For most predefined printers this will be unnecessary.

The plugin adds itself as supported action to all printers, except for printers with multiextrusion because they are not currently handled well. After this PR (which is a work in progress but should be ready before 2.6 goes into feature freeze), the Machine Settings action will be supported by any printer.

**Comment by fieldOfView** [comment]
> I would love it, if extruders in the side bar are greyed out to show that they are not being used.

Please log a separate issue for that, because that is well outside the scope of this PR.

**Comment by Drayson80** [comment]
Hmm.. how could I test/try this? Simply replace the files in my Cura plugin folder or do I have to do something else? (Pls. remind, I´m no software guy, just a mechanical one :-) )

**Comment by fieldOfView** [comment]
> (Pls. remind, I´m no software guy, just a mechanical one :-) )

Then I would say: don't bother yet. It is still very much a work in progress and not functional yet.

**Comment by Drayson80** [comment]
ok, thanks for the info. 
just so annoyed that I always need to switch printers when jumping between single and dual setup...

**Comment by fieldOfView** [comment]
Even with this PR, it is still going to be quicker to switch to another printer using the dropdown menu than it will be to go to the Manage Printers screen, click Machine Settings and then change the number of extruders.

**Comment by maukcc** [comment]
I see extruder offset is also being addressed in this?
If so, then #1279 can be closed.

**Comment by fieldOfView** [comment]
Yes, the x & y offset will be settable per extruder

**Comment by fieldOfView** [comment]
I have tentatively removed the [WIP] from the title of the PR. This can do with an initial round of review & testing

**Comment by fieldOfView** [comment]
Is Jenkins misbehaving again, or have I screwed up?

**Comment by awhiemstra** [comment]
Actually, it seems the network is misbehaving because Jenkins fails during a git checkout... :S

**Comment by Ghostkeeper** [comment]
See CURA-3299.

**Comment by nallath** [comment]
I'm kind of surprised about how little python code it requires for this to work.

**Comment by fieldOfView** [comment]
Testament of a well-thought-out settings system?

I did try a couple more Python-heavy approaches before ending up with this one.

**Comment by nallath** [comment]
So far I've managed to find a minor issue; Updating the offset does not trigger an update of the disallowed areas.

**Comment by nallath** [comment]
Another thing is that the project reader doesn't like definition_changes, but that is an issue of the project reader.

**Comment by fieldOfView** [comment]
I have changed the location where definition_changes containers are stored, because storing them in the folder with the stacks was ill-advised; the upgrade-system was trying to upgrade the definition-changes instance containers to container stacks.

File: plugins/MachineSettingsAction/MachineSettingsAction.qml [hunk 1]
```diff
@@ -671,7 +671,12 @@ Cura.MachineAction
                         propertyProvider.setPropertyValue("value", text);
                         if(forceUpdateOnChange)
                         {
+                            var extruderIndex = ExtruderManager.activeExtruderIndex;
                             manager.forceUpdate();
+                            if(ExtruderManager.activeExtruderIndex != extruderIndex)
+                            {
+                                ExtruderManager.setActiveExtruderIndex(extruderIndex)
+                            }
```
Comment (Ghostkeeper) [inline plugins/MachineSettingsAction/MachineSettingsAction.qml]: Hmm, the `forceUpdate` emits a `globalContainerStackChanged` signal. Does that change the active extruder index?

**Comment by jackha** [comment]
@fieldOfView: Wow this looks really amazing! I'm currently testing the PR and I wonder if you have the problem as well: it doesn't slice with this machine. Not even with 1 extruder. The progress bar is stuck at 0. Maybe some settings are missing or misplaced?

Error in CuraEngine (with debug mode): /Projects/CuraEngine/src/MeshGroup.cpp:77: cura::ExtruderTrain* cura::MeshGroup::getExtruderTrain(unsigned int): Assertion `extruders[extruder_nr]' failed.
2017-04-18 14:18:28,403 - DEBUG - CuraEngineBackend.CuraEngineBackend._onBackendQuit [545]: Backend quit with return code -6. Resetting process and socket.

File: cura/CuraApplication.py [hunk 2]
```diff
@@ -145,13 +146,15 @@ def __init__(self):
         Resources.addStorageType(self.ResourceTypes.UserInstanceContainer, "user")
         Resources.addStorageType(self.ResourceTypes.ExtruderStack, "extruders")
         Resources.addStorageType(self.ResourceTypes.MachineStack, "machine_instances")
+        Resources.addStorageType(self.ResourceTypes.DefinitionChangesContainer, "machine_settings")
```
Comment (awhiemstra) [inline cura/CuraApplication.py]: Just a minor comment, but I would prefer we use "definition_changes" for this. We should match the storage name to our internal naming as much as possible in my opinion.

File: plugins/MachineSettingsAction/MachineSettingsAction.qml (hunk 1 above)
Comment (fieldOfView) [inline plugins/MachineSettingsAction/MachineSettingsAction.qml]: Yes, globalContainerStackChanged seems to always set the active extruder index to 0, hence this fix.

File: cura/CuraApplication.py (hunk 2 above)
Comment (fieldOfView) [inline cura/CuraApplication.py]: Ok, point taken. I wanted to keep the folder close to "machine_instance".

**Comment by fieldOfView** [comment]
@jackha, hmm, I must admit that I had not tried slicing, mostly because slicing-wise I had already tested the base technique before. I will see what is going on.

**Comment by jackha** [comment]
@fieldOfView Thanks!

**Comment by BagelOrb** [comment]
It shouldn't matter on which it relies. The machine_extruder_count should
be exactly the same as the amount (count) of extruders. Isn't that obvious?

Op 18 apr. 2017 17:13 schreef "Aldo Hoeben" <notifications@github.com>:

It seems that somewhere CuraEngine is relying on the number of defined
extruders (metadata.machine_extruder_trains.length) instead of
machine_extruder_count. I'm still trying to find out where by staring at
the code. @BagelOrb <https://github.com/BagelOrb>, you can probably find
this faster than I can.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/Ultimaker/Cura/pull/1521#issuecomment-294876777>, or mute
the thread

**Comment by maukcc** [comment]
I am seeing 2 different things:
- the max amount of extruders possible on the machine. (in the definition .def.json file made by the vendor)
- the amount of extruders actually present on the machine (selected in the machine settings by the user)

**Comment by Ghostkeeper** [comment]
The problem with duplicate data is that the data might become inconsistent and this could lead to unexpected results. This is a textbook example! The fix is to depend on only one of the duplicates. In this case, I'd say depend on the extruder list.

There's quite a lot of things in the front-end that also depend on machine_extruder_count.

**Comment by BagelOrb** [comment]
You misunderstood. The machine setting should not be settable by the user.
The format doesn't allow you to make a distinction between possible
extruders and actual extruders.

You'll just manually have to make sure you only use the extruders currently
present on your machine.

Note that what is an extruder in cura actually governs a whole extruder
train: the extruder, the filament, the nozzle etc.

Could you send some pictures of your machine?! I have no idea how there are
multiple extruders defined but not present.

Op 19 apr. 2017 09:43 schreef "MaukCC" <notifications@github.com>:

I am seeing 2 different things:

   - the max amount of extruders possible on the machine. (in the
   definition .def.json file made by the vendor)
   - the amount of extruders actually present on the machine (selected in
   the machine settings by the user)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/Ultimaker/Cura/pull/1521#issuecomment-295143582>, or mute
the thread

**Comment by BagelOrb** [comment]
Is hard to find out the extruder count by looking at the number of
extruders defined because of inheritance.

A defe.json file can override a single extruder and leave the rest intact.

Op 19 apr. 2017 09:52 schreef "Tim Kuipers" <t.kuipers@ultimaker.com>:

**Comment by fieldOfView** [comment]
This PR tries to make it possible to define a custom printer with a settable number of extruders. Because of the back- and forward-references between machine definitions and extruder definitions, this is only possible by referencing a maximum number of extruders in the custom printer definition, and then setting the number of extruders to use in the machine instance.

I tried this method by creating a single extruder version of the UM3 using inheritance (which seems to work): https://gist.github.com/fieldOfView/bde7189087721451784865a28f3fc2c1
Unfortunately doing the same trick with an instance_container turns out not to work (yet).

**Comment by maukcc** [comment]
> this is only possible by referencing a maximum number of extruders in the custom printer definition, and then setting the number of extruders to use in the machine instance.

exactly what I mean.
This way I only have to make 1 cartesio.def.json file and then, on 1st start up, the user says how many tools are mounted to it. 
Other wise I would have to make 12 cartesio.def.json files (and its accompanying extruders, variants, materials, quality and so on), which will also probably slow down the startup time of Cura.

So at present when you select Cartesio printer at 1st start up then you get 4 extruders. Which works(ish) for the moment, even if there are only 2 extruders present.

**Comment by fieldOfView** [comment]
Is there an easy way to get a stacktrace out of CuraEngine?

**Comment by nallath** [comment]
The only way i know is with gdb.

**Comment by fieldOfView** [comment]
@jackha: this CuraEngine patch fixes the crash for me: https://github.com/Ultimaker/CuraEngine/pull/501
Would be great if you could test this too.

**Comment by fieldOfView** [comment]
I think there is still some work to do when the user sets the number of extruders to eg 4, then sets the support extruder to the 4th extruder, and then sets the number of extruders back to 2. Same with the other settings where you set an extruder_nr.

**Comment by Ghostkeeper** [comment]
I don't know if it's related to this change, but when I unfocus one of the machine head polygon text fields, I'm getting the log message:

    file:///home/ruben/Projects/Cura/plugins/MachineSettingsAction/MachineSettingsAction.qml:644: ReferenceError: printheadXMinField is not defined

It seems that printheadXMinField is out of scope for the `setHeadPolygon()` function, even though that very text field triggers the call to `setHeadPolygon()`. This error then prevents the head polygon from being updated and makes these text fields ineffective.

**Comment by Ghostkeeper** [comment]
When I switch to one of the extruder tabs I'm getting these error messages:

    file:///usr/local/lib/python3.5/dist-packages/PyQt5/Qt/qml/QtQuick/Controls/ScrollView.qml:348:13: QML Item: Possible anchor loop detected on fill.
    file:///usr/local/lib/python3.5/dist-packages/PyQt5/Qt/qml/QtQuick/Controls/ScrollView.qml:348:13: QML Item: Possible anchor loop detected on fill.
    file:///home/ruben/Projects/Cura/plugins/MachineSettingsAction/MachineSettingsAction.qml:665:23: Unable to assign [undefined] to QString
    file:///home/ruben/Projects/Cura/plugins/MachineSettingsAction/MachineSettingsAction.qml:665:23: Unable to assign [undefined] to QString
    file:///home/ruben/Projects/Cura/plugins/MachineSettingsAction/MachineSettingsAction.qml:665:23: Unable to assign [undefined] to QString
    file:///home/ruben/Projects/Cura/plugins/MachineSettingsAction/MachineSettingsAction.qml:585:43: Unable to assign [undefined] to QString
    file:///home/ruben/Projects/Cura/plugins/MachineSettingsAction/MachineSettingsAction.qml:609:43: Unable to assign [undefined] to QString

**Comment by Ghostkeeper** [comment]
And indeed with the other patch, it slices fine. At least, I did a test with an F model that had support on the second extruder and that worked.

When I set the support pattern to be triangles though, nothing happens. It stays on zigzag. This indicates to me that there is some internal inconsistency still.

When that has been fixed and the error when switching extruder tabs in the machine action window is fixed too, I deem this pull request ready to merge (along with the accompanying pull request in the engine).

**Comment by fieldOfView** [comment]
Thanks for the review. Will look in to the issues you found.

**Comment by fieldOfView** [comment]
> When I set the support pattern to be triangles though, nothing happens. It stays on zigzag. 

I think this is caused by a setting value on the relevant extruder that is being overridden by a value that is still set on the global stack. I will see if I can fix this by moving non-global-only settings to the first extruder stack when switching to multiple extruders (and moving the same settings back to the global stack when switching back to single extruders).

**Comment by fieldOfView** [comment]
Ready for another round of review/tests. I fixed the issues found in the previous round, and fixed a number of issues when switching the number of extruders. The definition_changes containers are now also properly disposed of.

**Comment by fieldOfView** [comment]
> All checks have failed

Not sure if me or Jenkins

**Comment by nallath** [comment]
[WS-CLEANUP] Deleting project workspace...Cannot delete workspace :java.io.EOFException 

Jenkins...

**Comment by jackha** [comment]
@fieldOfView: The support pattern in multi extruder machine is still a problem. Still stays on zig-zag. We think that it's important to fix this, as it may affect other settings as well.

btw. The CuraEngine PR is merged.

**Comment by fieldOfView** [comment]
Could you add reproduction steps? I could not reproduce it. Preferably starting from a clean printer.

**Comment by jackha** [comment]
I used the feature_multiextruder_machinesettings with current master merged into it + master CuraEngine + master Uranium. Then, when I start clean, add fdm printer with 2 extruders it doesn't use the Support Pattern.

**Comment by fieldOfView** [comment]
Ah, I see now. The support pattern only works if the number of extruders is 1 or 8, but not anywhere inbetween. Weird. I think I'll have to dive into CuraEngine again.

**Comment by fieldOfView** [comment]
My CuraEngine fix was the culprit. It could ignore all extruder settings. 
See https://github.com/Ultimaker/CuraEngine/pull/509 for the fix.

**Comment by jackha** [comment]
Ok merged it into master after fixing a small problem. 

Strange that github says it has conflicts... oh well it's merged anyway.

**Comment by jackha** [comment]
btw Did I say Thanks for the big effort! ?!

**Comment by maukcc** [comment]
I am loving it. 
I am having a little mishap: 
I have 4 extruder trains in the definition file, but I choose to use only 2 in my machine settings.
Now Cura lets me see the 2 extruders, but takes the other 2 extruders(which are not there) into account.
For instance I have quality profiles where it is not allowed to make layerheights heigher then 0.3mm if one of the nozzles is smaller then 0.5mm.
Now if one of the non visible extruder trains is set to 0.4, it will not show these quality profiles.

**Comment by fieldOfView** [comment]
> For instance I have quality profiles where it is not allowed to make layerheights higher then 0.3mm if one of the nozzles is smaller then 0.5mm.

Silly question: how do you do that?
In other words: how can I reproduce this?

**Comment by maukcc** [comment]
You can select the Cartesio as machine and try it.

In the definition file we set:
"layer_height": { "maximum_value": "(0.8 * min(extruderValues('machine_nozzle_size')))" },
and then it will not show the quality profiles with a layerheight above that min value.

And that is also the culprit as that line checks all extruderValues/trains and not just the ones actually in use.

**Comment by maukcc** [comment]
No that was not it.
It is the ""has_machine_quality": true," line in the definition file that does the trick.

**Comment by fieldOfView** [comment]
I am working on a patch. Are you running Cura "from source"? If so you could help test the patch.

**Comment by maukcc** [comment]
No sorry, I am running thopiekar`s master version.

**Comment by fieldOfView** [comment]
@maukcc, see #1809

**Comment by maukcc** [comment]
With today’s version of thopiekar`s master version I can:

change extruder amount, but it will not update the amount
So in settings I can change the amount of extruders, ig from 4 to 2, but it will keep showing the 4 extruder tabs. Until I change the machine and then change it back again, then the value has been activated and it only shows the 2 tabs.(or visa versa)

values that are set for unselected extruders are impacting
So if I set extruder_3 to 0.4 nozzle, and I say that the machine has only 2 extruders, It will still read the value of extruder_3, and I can not use larger layer heights.

**Comment by fieldOfView** [comment]
Please start a new issue. This was working, and now it is broken.

**Pull request closed without merging**